Punctuation Checker

benjaminking opened the pull request:
I think I have this rebased properly, but the list of changes looks a little unusual (it makes it look like I made changes to some of the files in "core").  If there's something I need to differently, let me know.

<!-- Reviewable:start -->
- - -
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/sillsdev/lynx/12)
<!-- Reviewable:end -->

benjaminking commented:
Ok, merging main into the branch seems like it fixed things.  Thanks!

ddaspit reviewed (changes requested):
This is really good work. The code is well organized and extensible. The names are very clear. It has good typing and clean Typescript. The unit tests are very extensive.

> Reviewed 70 of 91 files at r1, 21 of 21 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/lynx/12#-:-OJLAFd14oGBvwogYhB5:b-974acq)* status: all files reviewed, 7 unresolved discussions (waiting on @benjaminking)

___
*[`packages/punctuation-checker/tsconfig.json` line 5 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJKnxF703A9Y7AxDNMy:-OJKnxF703A9Y7AxDNMz:b-yil1rq) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/packages/punctuation-checker/tsconfig.json#L5)):*
> ```jsonc
>   "compilerOptions": {
>     "outDir": "dist",
>     "rootDir": "."
> ```

The `rootDir` should be set to `src`.
___
*[`packages/punctuation-checker/package.json` line 3 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJKk7oTCPBHD4housWq:-OJKk7oTCPBHD4housWr:b67xzif) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/packages/punctuation-checker/package.json#L3)):*
> ```JSON
> {
>   "name": "@sillsdev/lynx-punctuation-checker",
>   "version": "1.0.0",
> ```

You should change the version to `0.0.0`.
___
*[`packages/punctuation-checker/package.json` line 9 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJKnUZrB-TdN-jS3ShW:-OJKnUZrB-TdN-jS3ShX:b-ht6asd) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/packages/punctuation-checker/package.json#L9)):*
> ```JSON
>   "main": "./dist/index.cjs",
>   "module": "./dist/index.js",
>   "exports": {
> ```

I suspect there might be some issues with the way that the Typescript types are setup. Can you run [arethetypeswrong](https://github.com/arethetypeswrong/arethetypeswrong.github.io/blob/main/packages/cli/README.md) to ensure that there aren't any issues? 
___
*[`packages/punctuation-checker/package.json` line 17 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJKlGl5A14xgYrNa9B3:-OJKlGl5A14xgYrNa9B4:bqxxvig) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/packages/punctuation-checker/package.json#L17)):*
> ```JSON
>   },
>   "scripts": {
>     "build": "tsup src/index.ts --dts --format cjs,esm --clean --sourcemap",
> ```

You can use the `tsup-config` internal package to configure tsup. See the other packages for examples.
___
*[`packages/punctuation-checker/package.json` line 28 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJKkTjyBxkYVTiwIXaM:-OJKkTjyBxkYVTiwIXaN:b-t0amh0) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/packages/punctuation-checker/package.json#L28)):*
> ```JSON
>   "dependencies": {
>     "rxjs": "^7.8.1",
>     "@sillsdev/lynx": "*",
> ```

You should reference the current version of the Lynx libraries, i.e. `^0.3.0` for `@sillsdev/lynx`.
___
*[`package.json` line 30 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJKsGn9D5YbLQLmY004:-OJKsGn9D5YbLQLmY005:by9udev) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/package.json#L30)):*
> ```JSON
>     "turbo": "^2.1.3",
>     "vitest": "^2.1.5",
>     "typescript": "^5.5.4",
> ```

Can you tell me more about the added dependencies and what they are needed for? If they are needed by the punctuation checker, then the dependencies should be added in that package.
___
*[`packages/punctuation-checker/src/abstract-checker.ts` line 26 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJL6ulh1tzSvnop2R0U:-OJL6ulh1tzSvnop2R0V:b-o3f7kq) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/packages/punctuation-checker/src/abstract-checker.ts#L26)):*
> ```XML
>     private readonly issueFinderFactory: IssueFinderFactory,
>   ) {
>     this.diagnosticsChanged$ = merge(
> ```

Scripture Forge uses the `diagnosticsChanged$` observable to listen for diagnostics. It looks like right now you are returning diagnostics for the currently open document. I think this is fine right now, but in SF, we will probably want to show diagnostics for other documents as well. SF has the ability to show issues in the rest of translation and not just the currently open chapter. I don't think we have a good sense of how much of an impact validating all chapters of a translation will have on performance. I have set up SF to only fire change events for the currently open chapter, so we won't need to validate every chapter every time they change. If it is too much of an impact to validate all chapters, we could validate all of the chapters in the currently open book instead. We will have to see. I just wanted to let you know that we might need to update the punctuation checker to validate more than just the currently open chapter.
___
*[`packages/punctuation-checker/src/allowed-character/allowed-character-checker.ts` line 38 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJL1a9G2C-Vb0VHsjFK:-OJL1a9H91AroO7gXrqi:b-2j8g3w) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/packages/punctuation-checker/src/allowed-character/allowed-character-checker.ts#L38)):*
> ```XML
>       this.localizer.addNamespace(
>         ALLOWED_CHARACTER_CHECKER_LOCALIZER_NAMESPACE,
>         (language: string) => import(`./locales/${language}.json`, { with: { type: 'json' } }),
> ```

I get a spurious warning from vite when running tests. I suppressed this warning in the `examples` package by adding a `vitest.config.js` file that sets the log level to `error`.


<!-- Sent from Reviewable.io -->

ddaspit reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/sillsdev/lynx/12)* status: all files reviewed, 8 unresolved discussions (waiting on @benjaminking)

___
*[`packages/punctuation-checker/src/allowed-character/allowed-character-checker.ts` line 35 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJLe0cQ41rIK3TNOQ_2:-OJLe0cQ41rIK3TNOQ_3:b-dnpuen) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/packages/punctuation-checker/src/allowed-character/allowed-character-checker.ts#L35)):*
> ```XML
>     // tests can provide different messages, but due to the way variable dynamic imports
>     // work, the namespace loading function can only appear in this file at this location
>     if (!this.localizer.hasNamespace(ALLOWED_CHARACTER_CHECKER_LOCALIZER_NAMESPACE)) {
> ```

The `addNamespace` method already checks if the namespace has been added. Did you encounter a problem with the `addNamespace` method that caused you to add the `hasNamespace`method?


<!-- Sent from Reviewable.io -->

benjaminking commented:
To save space, I'm skipping the items that I've fixed, but don't have a comment for.

> _[`packages/punctuation-checker/tsconfig.json` line 5 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJKnxF703A9Y7AxDNMy:-OJKnxF703A9Y7AxDNMz:b-yil1rq) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/packages/punctuation-checker/tsconfig.json#L5)):_
> 
> > ```js
> >   "compilerOptions": {
> >     "outDir": "dist",
> >     "rootDir": "."
> > ```
> 
> The `rootDir` should be set to `src`.

I ended up creating two different tsconfig files, the main one with `rootDir` set to `src`, and a `tsconfig.dev.json` that has `rootDir` set to `.` for the sake of the tests folder.  I'm not sure if that's the preferred solution, but it was able to pass all the checks.

> _[`package.json` line 30 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJKsGn9D5YbLQLmY004:-OJKsGn9D5YbLQLmY005:by9udev) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/package.json#L30)):_
> 
> > ```json
> >     "turbo": "^2.1.3",
> >     "vitest": "^2.1.5",
> >     "typescript": "^5.5.4",
> > ```
> 
> Can you tell me more about the added dependencies and what they are needed for? If they are needed by the punctuation checker, then the dependencies should be added in that package.

I think these were required for me to build/test correctly with some earlier configuration of Lynx before we settled on the current tooling and setup.  I've removed these requirements and haven't encountered the errors again.



> _[`packages/punctuation-checker/src/allowed-character/allowed-character-checker.ts` line 35 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJLe0cQ41rIK3TNOQ_2:-OJLe0cQ41rIK3TNOQ_3:b-dnpuen) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/packages/punctuation-checker/src/allowed-character/allowed-character-checker.ts#L35)):_
> 
> > ```
> >     // tests can provide different messages, but due to the way variable dynamic imports
> >     // work, the namespace loading function can only appear in this file at this location
> >     if (!this.localizer.hasNamespace(ALLOWED_CHARACTER_CHECKER_LOCALIZER_NAMESPACE)) {
> > ```
> 
> The `addNamespace` method already checks if the namespace has been added. Did you encounter a problem with the `addNamespace` method that caused you to add the `hasNamespace`method?

The check in `addNamespace` seems to work differently.  For some reason, `this.backend.hasNamespace` and `this.i18n.hasLoadedNamespace` return different results.  As an example, the test at `allowed-character-checker.test.ts:47` will fail if I remove the `hasNamespace` check in `allowed-character-checker.ts` -- the test's messages are overwritten by the messages in `locales/en.json`.

(By the way, the reason for this awkward setup is that variable dynamic imports are disallowed outside of the folder containing `locales`.  This means I can't load the default localization messages until after I've already injected the localizer. So I have to check whether messages have been loaded already, so that I don't overwrite any messages that the tests might have loaded beforehand.)

ddaspit reviewed (commented):
> Reviewed 7 of 7 files at r3, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/lynx/12)* status: all files reviewed, 3 unresolved discussions (waiting on @benjaminking)

___
*[`packages/punctuation-checker/package.json` line 17 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJKlGl5A14xgYrNa9B3:-OJZZrhh5iE36Qu-EvGi:b-wwbyam) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/packages/punctuation-checker/package.json#L17)):*
<details><summary><i>Previously, ddaspit (Damien Daspit) wrote…</i></summary><blockquote>

You can use the `tsup-config` internal package to configure tsup. See the other packages for examples.
</blockquote></details>

You should add a dev dependency on `@repo/tsup-config`.
___
*[`packages/punctuation-checker/tsconfig.json` line 5 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJKnxF703A9Y7AxDNMy:-OJZ_I371scjknUuY2jL:b-vqmsy7) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/packages/punctuation-checker/tsconfig.json#L5)):*
<details><summary><i>Previously, benjaminking (Ben King) wrote…</i></summary><blockquote>

I ended up creating two different tsconfig files, the main one with `rootDir` set to `src`, and a `tsconfig.dev.json` that has `rootDir` set to `.` for the sake of the tests folder.  I'm not sure if that's the preferred solution, but it was able to pass all the checks.
</blockquote></details>

I see. The original `tsconfig.json` is fine. You can undo this change.
___
*[`packages/punctuation-checker/src/allowed-character/allowed-character-checker.ts` line 35 at r2](https://reviewable.io/reviews/sillsdev/lynx/12#-OJLe0cQ41rIK3TNOQ_2:-OJZorE3EyrmeX46CaCJ:b1s33j0) ([raw file](https://github.com/sillsdev/lynx/blob/b741aa407bb21adc35ccc269f35704bc64fa906c/packages/punctuation-checker/src/allowed-character/allowed-character-checker.ts#L35)):*
<details><summary><i>Previously, benjaminking (Ben King) wrote…</i></summary><blockquote>

The check in `addNamespace` seems to work differently.  For some reason, `this.backend.hasNamespace` and `this.i18n.hasLoadedNamespace` return different results.  As an example, the test at `allowed-character-checker.test.ts:47` will fail if I remove the `hasNamespace` check in `allowed-character-checker.ts` -- the test's messages are overwritten by the messages in `locales/en.json`.

(By the way, the reason for this awkward setup is that variable dynamic imports are disallowed outside of the folder containing `locales`.  This means I can't load the default localization messages until after I've already injected the localizer. So I have to check whether messages have been loaded already, so that I don't overwrite any messages that the tests might have loaded beforehand.)

</blockquote></details>

Could we update the `addNamespace` method to use `this.backend.hasNamespace` instead of `this.i18n.hasLoadedNamespace`?


<!-- Sent from Reviewable.io -->

benjaminking commented:
As I was looking through the tsup configuration, I realized that my `locales` folders are not being copied to `dist`.  This is because my `locales` folders are in a different location than where the `library.js` script is looking for them.  One solution I was considering was to modify the command in `library.js` to copy from multiple potential locations, e.g.:

`for location in . allowed-character quotation [...] ; do if [ -d src/$location/locales ]; then copy-folder src/$location/locales dist/$location/locales; fi; done`

or

`find src -name "locales" -type d -printf '%P\n' | while read dir; do copy-folder src/$dir dist/$dir; done`

Alternatively, I could combine all the locale files into a single file in the `src/locales`, where the tsup script is already expecting to find them.  But that would require significant refactoring, due to the restrictions on which files are allowed to perform variable dynamic imports.

What are your thoughts on this?

ddaspit commented:
Good catch. I think it makes sense to make `library.js` more generic with respect to the location of the `locales` folder.

benjaminking commented:
> Could we update the `addNamespace` method to use `this.backend.hasNamespace` instead of `this.i18n.hasLoadedNamespace`?

I made the switch to use `this.backend.hasNamespace` and removed the `hasNamespace` method (along with the calls to it).  All the tests passed, so it seems to do what we want.



> Good catch. I think it makes sense to make `library.js` more generic with respect to the location of the `locales` folder.

I also updated the command in `library.js` to search for folders named `locales` and copy them over to the respective location in `dist`.

ddaspit reviewed (commented):
You need to [add a changeset](https://github.com/changesets/changesets/blob/main/docs/adding-a-changeset.md) to the PR.

> Reviewed 12 of 12 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/lynx/12#-:-OJdTU5O7aEMRXF5C7Rx:b-gfhb4h)* status: all files reviewed, 1 unresolved discussion (waiting on @benjaminking)

___
*[`packages/tsup-config/library.js` line 11 at r4](https://reviewable.io/reviews/sillsdev/lynx/12#-OJdSH0K6MejCSBJdsfP:-OJdSH0K6MejCSBJdsfQ:b-u378yu) ([raw file](https://github.com/sillsdev/lynx/blob/53a105babca008c868b7be92b647715dd2739ccc/packages/tsup-config/library.js#L11)):*
> ```JavaScript
>       ...configOptions,
>       onSuccess:
>         'find src -name "locales" -type d -printf \'%P\n\' | while read dir; do copy-folder src/$dir dist/$dir; done',
> ```

This doesn't work on my Windows machine. You might be able to find another copy package that does what we need. Maybe [this package](https://www.npmjs.com/package/cpy-cli). `onSuccess` can also take a function, so you could implement the copy yourself.


<!-- Sent from Reviewable.io -->

benjaminking commented:
> This doesn't work on my Windows machine. You might be able to find another copy package that does what we need. Maybe [this package](https://www.npmjs.com/package/cpy-cli). `onSuccess` can also take a function, so you could implement the copy yourself.

I was able to get the new copy utility you mentioned working so that it correctly copies all the different file/folder structures for locales.

ddaspit reviewed (commented):
It looks like the `package-lock.json` needs to be updated to remove references to the `copy-folder-util` package. I think if you run `npm install`, it will update the lock file.

> Reviewed 3 of 3 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/lynx/12#-:-OJu3f6ACyIkYNKsxZAP:b-edv54i)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @benjaminking)



<!-- Sent from Reviewable.io -->

benjaminking commented:
> It looks like the `package-lock.json` needs to be updated to remove references to the `copy-folder-util` package. I think if you run `npm install`, it will update the lock file.

Ah, I had always assumed that `npm ci` included a run of `npm install` -- good to know that it does not.  The new `package-lock.json` should be pushed now.

ddaspit reviewed (commented):
Everything looks good. The last thing is to add a changeset using the Changesets CLI.

> Reviewed 1 of 1 files at r6, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/lynx/12#-:-OJy0geD1dw7RyjvuI5r:bq5qzwd)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @benjaminking)



<!-- Sent from Reviewable.io -->

benjaminking commented:
Ok.  I have added a changeset with the punctuation checker getting a minor version change and lynx-core getting a patch change.

ddaspit reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 1 of 1 files at r7, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/sillsdev/lynx/12#-:-OK32oaw2vpyfNrYWKji:bnfp4nl)* status: <span>:shipit:</span> complete! all files reviewed, all discussions resolved (waiting on @benjaminking)



<!-- Sent from Reviewable.io -->
